### ISSUE TITLE
Normalize latitude should work with latitude values greater than 180 degrees.

### ISSUE THREAD

**Issue by jrepko** [WorldWindEarth/WorldWindJava#46]
https://github.com/WorldWindEarth/WorldWindJava/blob/dcb04132c2daeafa91e347394ea78f5790ca3629/src/gov/nasa/worldwind/geom/Angle.java#L684

The normalize longitude method is doing the correct thing but the latitude method is not.

If you make an Angle with fromDMS that is meant to be in the southern hemisphere, you have to make it a positive number.  But when you call the normalize latitude method, you do not get your negative value back.   So, -11 degrees becomes 169 degrees, but then becomes 11, not -11 when you normalize.  

Maybe I'm missing something and someone could explain.

```
        int lat = -11;
        Angle angTest = Angle.fromDMS(180 + lat, 0, 0);
        angTest = angTest.normalizedLatitude();

        assertEquals("not working", "-11.0°", angTest.toString());
```

**Comment by wcmatthysen** [comment]
@jrepko, @jmmain69 and I had a discussion way back about this in [issue 25](https://github.com/NASAWorldWind/WorldWindJava/issues/25) of the upstream-repository. The result of this was that the `Angle.fromDMS()` method was changed to only accept positive numbers for the degrees parameter. The issue had to do with the signum function in the original method, as in:

    Math.signum(degrees) * (Math.abs(degrees) + minutes / 60d + seconds / 3600d)

will return 0, if degrees was less than 1 because `Math.signum(num)` will be 0 if `num` is less than 1.0. I proposed that we still tried to detect the sign for the degrees-parameter but if you look at the conversation there it was futile and my concerns was ignored. I was especially concerned about changing existing behaviour as the code used to always accept negative values for the degrees parameter. However, the beauty of the community fork is that we can now fix that bug here. Just keep in mind the subtleties of `Math.signum()` as noted in the original upstream issue.

**Comment by jmmain69** [comment]
The true problem is that even without signum() there is no way to send a -0 integer to the function to represent -0 degrees so any coordinate between -1 and 0 will be returned as a positive.  This is as unacceptable as returning 0 for all coordinates between -1 and 1 degree.

**Comment by wcmatthysen** [comment]
@jmmain69, you are correct. However, we could change the method signature to accept a floating point value for the degrees parameter. That way we could allow for minus-zero to be passed as a value for the degrees parameter. We then just round the degrees parameter and detect its sign (maybe without signum) and use that in the method body to do the calculation.

**Comment by jmmain69** [comment]
I’m not opposed to adding a new function with a different signature and documentation.  I was just pointing out that the current function as defined cannot cover the full spectrum of positive and negative values, and cannot be modified to cover the range of values between -1 and 0.

**Comment by wcmatthysen** [comment]
Jip, you are correct, the current function with integer signature won't work for degrees between -1 and 0. I just think that in the original change that the upstream guys made, they just went ahead and forced the degrees parameter to be positive even though the method always worked for negative degrees values. This could cause problems if you used WorldWind in a production environment where you depend on the method accepting negative values and suddenly the behaviour of this method just changed. If you didn't pick up that the method's behaviour changed and you upgraded to a newer version of WorldWind your code would suddenly just break. In my opinion the bug should have been fixed for 0-degrees, i.e. detect that the degrees value is zero and do not use signum for that case, as in:

    public static Angle fromDMS(int degrees, int minutes, int seconds)
    {
        ...
        float sign = (degrees == 0) ? 1.0f : Math.signum(degrees);
        return Angle.fromDegrees(sign * (Math.abs(degrees) + minutes / 60d + seconds / 3600d));
    }

Then, maybe introduce an additional method where the first parameter is a float so that people can pass a negative zero degrees value, as in:

    public static Angle fromDMS(float degrees, int minutes, int seconds)
    {
        ....
        float sign = (1 / degrees > 0) ? 1.0f : -1.0;
        return Angle.fromDegrees(sign * (Math.abs(Math.round(degrees)) + minutes / 60d + seconds / 3600d));
    }

This way you fix existing code and you expand the functionality to allow for negative zero degree values. You can then construct angles between -1 and 0 as follows:

    Angle angle = Angle.fromDMS(-0.0, 1, 1);

Or, if this is not elegant, we can drop the method with the float parameter and then just indicate to people that if they want to create angles between -1 and 0 degrees they need to do the following:

    Angle angle = Angle.fromDMS(0, 1, 1).multiply(-1);

This would then still work, but also still allow for negative degrees to be specified as in:

    Angle angle = Angle.fromDMS(-11, 0, 0);

**Comment by jmmain69** [comment]
It did cause problems where we used it in a production environment and it gave us bad values!

**Comment by jrepko** [comment]
@wcmatthysen i understand.  And that is definitely more readable than trying to undo the normalization to make a unit test, which is what I was trying to do.

Instead of trying to handle a -0 or handling negative values directly at all, you could make a fromDMS that takes 4 parameters.  Leave the first three params as is.  The last parameter could be an enum of N,S,E,W and let the enum indicate the provided direction.  It might make a little more sense than multiplying by -1 to get negative values.

**Comment by gbburkhardt** [comment]
I don't think that the 'fromDMS' method should be changed at all.  It's clear that it works only for positive values.  It's easy to adapt.  However, I'm curious as to what the desired answer is for 'normalizedDegreesLatitude'.  Should 184 deg be normalized to -4?  That would be my first guess.  But, of course, that's not the answer it currently gets.

**Comment by wcmatthysen** [comment]
@gbburkhardt, we can keep the `fromDMS()` method as is. I like the idea of @jrepko where we introduce an additional method taking 4 parameters with the direction as the last parameter.

I think you are correct in that 184-degrees should be normalized to -4-degrees (and not 4-degrees as is currently the case). This should probably be true for the negative case as well, where -184-degrees should be normalized to 4-degrees (instead of -4-degrees).

**Comment by Sufaev** [comment]
Just add hemisphere check:

public static double normalizedDegreesLatitude(double degrees) {
        double lat = degrees % 180;
        double normalizedLat = lat > 90 ? 180 - lat : lat < -90 ? -180 - lat : lat;
        // Determine whether whether the latitude is in the north or south hemisphere
        int numEquatorCrosses = (int) (degrees / 180);
        return (numEquatorCrosses % 2 == 0) ? normalizedLat : -normalizedLat;
    }

**Comment by wcmatthysen** [comment]
@jrepko, @gbburkhardt, @Sufaev, I made the fix in PR #63. Check if you guys are happy with this and we can merge into develop.

**Comment by Sufaev** [comment]
I am ok with this solution.

**Comment by PJHogan** [comment]
I'm very happy with it!

**Comment by wcmatthysen** [comment]
Merged in PR #63.
